File: packages/sdk-python/rollgate/cache.py
"""
Flag cache with stale-while-revalidate support.
"""

import json
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional, Callable, Dict, Any


@dataclass
class CacheConfig:
    """Configuration for cache behavior."""

    ttl_ms: int = 300000
    """Time-to-live for fresh cache entries (default: 5 minutes)."""

    stale_ttl_ms: int = 3600000
    """Time-to-live for stale cache entries (default: 1 hour)."""

    persist_path: Optional[str] = None
    """File path for persistent cache."""


@dataclass
class CacheStats:
    """Cache statistics."""

    hits: int = 0
    misses: int = 0
    stale_hits: int = 0
    size: int = 0


@dataclass
class CacheEntry:
    """Cache entry with metadata."""

    value: Dict[str, bool]
    timestamp: float
    stale: bool = False


@dataclass
class CacheResult:
    """Result of a cache lookup."""

    flags: Dict[str, bool]
    stale: bool


DEFAULT_CACHE_CONFIG = CacheConfig()


class FlagCache:
    """
    Flag cache with stale fallback support.

    Features:
    - In-memory caching with configurable TTL
    - Stale-while-revalidate pattern
    - File persistence
    - Event callbacks for cache state changes
    """

    def __init__(self, config: Optional[CacheConfig] = None):
        self._config = config or DEFAULT_CACHE_CONFIG
        self._cache: Dict[str, CacheEntry] = {}
        self._stats = CacheStats()
        self._callbacks: Dict[str, list[Callable]] = {
            "cache_hit": [],
            "cache_miss": [],
            "cache_set": [],
            "cache_expired": [],
            "cache_stale": [],
        }

    def on(self, event: str, callback: Callable) -> None:
        """Register an event callback."""
        if event in self._callbacks:
            self._callbacks[event].append(callback)

    def off(self, event: str, callback: Callable) -> None:
        """Remove an event callback."""
        if event in self._callbacks and callback in self._callbacks[event]:
            self._callbacks[event].remove(callback)

    def _emit(self, event: str, *args) -> None:
        """Emit an event to all registered callbacks."""
        for callback in self._callbacks.get(event, []):
            try:
                callback(*args)
            except Exception:
                pass

    def get(self, key: str = "flags") -> Optional[CacheResult]:
        """
        Get cached flags.

        Args:
            key: Cache key

        Returns:
            CacheResult with flags and stale indicator, or None if not found
        """
        entry = self._cache.get(key)

        if entry is None:
            self._stats.misses += 1
            self._emit("cache_miss", key)
            return None

        age_ms = (time.time() - entry.timestamp) * 1000

        # Fresh cache
        if age_ms < self._config.ttl_ms:
            self._stats.hits += 1
            self._emit("cache_hit", key, False, age_ms)
            return CacheResult(flags=entry.value, stale=False)

        # Stale but usable
        if age_ms < self._config.stale_ttl_ms:
            self._stats.stale_hits += 1
            self._emit("cache_hit", key, True, age_ms)
            self._emit("cache_stale", key, age_ms)
            return CacheResult(flags=entry.value, stale=True)

        # Expired - remove from cache
        del self._cache[key]
        self._stats.size = len(self._cache)
        self._stats.misses += 1
        self._emit("cache_expired", key, age_ms)
        return None

    def set(self, key: str, flags: Dict[str, bool]) -> None:
        """
        Store flags in cache.

        Args:
            key: Cache key
            flags: Flags dictionary
        """
        entry = CacheEntry(
            value=flags,
            timestamp=time.time(),
            stale=False,
        )

        self._cache[key] = entry
        self._stats.size = len(self._cache)
        self._emit("cache_set", key, len(flags))

        # Persist if configured
        self._persist()

    def has_fresh(self, key: str = "flags") -> bool:
        """Check if cache has fresh data."""
        entry = self._cache.get(key)
        if entry is None:
            return False
        age_ms = (time.time() - entry.timestamp) * 1000
        return age_ms < self._config.ttl_ms

    def has_any(self, key: str = "flags") -> bool:
        """Check if cache has any data (fresh or stale)."""
        entry = self._cache.get(key)
        if entry is None:
            return False
        age_ms = (time.time() - entry.timestamp) * 1000
        return age_ms < self._config.stale_ttl_ms

    def clear(self) -> None:
        """Clear all cached data."""
        self._cache.clear()
        self._stats.size = 0

    def get_stats(self) -> CacheStats:
        """Get cache statistics."""
        return CacheStats(
            hits=self._stats.hits,
            misses=self._stats.misses,
            stale_hits=self._stats.stale_hits,
            size=self._stats.size,
        )

    def get_hit_rate(self) -> float:
        """Get hit rate (hits / (hits + misses))."""
        total = self._stats.hits + self._stats.stale_hits + self._stats.misses
        if total == 0:
            return 0.0
        return (self._stats.hits + self._stats.stale_hits) / total

    def load(self) -> bool:
        """
        Load cache from persistent storage.

        Returns:
            True if cache was loaded successfully
        """
        if not self._config.persist_path:
            return False

        try:
            path = Path(self._config.persist_path)
            if not path.exists():
                return False

            data = json.loads(path.read_text())
            entries = data.get("entries", [])

            for key, entry_data in entries:
                timestamp = entry_data.get("timestamp", 0)
                age_ms = (time.time() - timestamp) * 1000

                # Only restore if within stale TTL
                if age_ms < self._config.stale_ttl_ms:
                    self._cache[key] = CacheEntry(
                        value=entry_data.get("value", {}),
                        timestamp=timestamp,
                        stale=age_ms >= self._config.ttl_ms,
                    )

            self._stats.size = len(self._cache)
            return True
        except Exception:
            return False

    def _persist(self) -> bool:
        """
        Persist cache to storage.

        Returns:
            True if cache was persisted successfully
        """
        if not self._config.persist_path:
            return False

        try:
            data = {
                "version": 1,
                "entries": [
                    [key, {"value": entry.value, "timestamp": entry.timestamp}]
                    for key, entry in self._cache.items()
                ],
            }
            path = Path(self._config.persist_path)
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(json.dumps(data))
            return True
        except Exception:
            return False

    def close(self) -> None:
        """Cleanup resources and final persist."""
        if self._config.persist_path:
            self._persist()

File: packages/sdk-core/src/cache.ts
/**
 * Simple EventEmitter for browser compatibility
 */
type EventCallback = (...args: unknown[]) => void;

class SimpleEventEmitter {
  private listeners: Map<string, Set<EventCallback>> = new Map();

  on(event: string, callback: EventCallback): void {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(callback);
  }

  off(event: string, callback: EventCallback): void {
    this.listeners.get(event)?.delete(callback);
  }

  emit(event: string, ...args: unknown[]): void {
    this.listeners.get(event)?.forEach((cb) => cb(...args));
  }

  removeAllListeners(): void {
    this.listeners.clear();
  }
}

/**
 * Cache configuration
 */
export interface CacheConfig {
  /** Time-to-live for fresh cache entries (default: 300000ms = 5min) */
  ttl: number;
  /** Time-to-live for stale cache entries (default: 3600000ms = 1h) */
  staleTtl: number;
  /** localStorage key for persistence (optional) */
  storageKey?: string;
}

export const DEFAULT_CACHE_CONFIG: CacheConfig = {
  ttl: 300000, // 5 minutes
  staleTtl: 3600000, // 1 hour
};

/**
 * Cache entry with metadata
 */
interface CacheEntry {
  flags: Record<string, boolean>;
  timestamp: number;
}

/**
 * Cache statistics
 */
export interface CacheStats {
  hits: number;
  misses: number;
  staleHits: number;
}

/**
 * Flag cache with stale fallback support
 *
 * Features:
 * - In-memory caching with configurable TTL
 * - Stale-while-revalidate pattern
 * - localStorage persistence for browser
 * - Event emission for cache hit/miss/stale
 */
export class FlagCache extends SimpleEventEmitter {
  private cache: CacheEntry | null = null;
  private config: CacheConfig;
  private stats: CacheStats = {
    hits: 0,
    misses: 0,
    staleHits: 0,
  };

  constructor(config: Partial<CacheConfig> = {}) {
    super();
    this.config = { ...DEFAULT_CACHE_CONFIG, ...config };
  }

  /**
   * Get cached flags
   */
  get(): { flags: Record<string, boolean>; stale: boolean } | undefined {
    if (!this.cache) {
      this.stats.misses++;
      this.emit("cache-miss", { key: "flags" });
      return undefined;
    }

    const age = Date.now() - this.cache.timestamp;

    // Fresh cache
    if (age < this.config.ttl) {
      this.stats.hits++;
      this.emit("cache-hit", { key: "flags", age });
      return { flags: this.cache.flags, stale: false };
    }

    // Stale but usable
    if (age < this.config.staleTtl) {
      this.stats.staleHits++;
      this.emit("cache-stale", { key: "flags", age });
      return { flags: this.cache.flags, stale: true };
    }

    // Expired - clear cache
    this.cache = null;
    this.stats.misses++;
    this.emit("cache-miss", { key: "flags" });
    return undefined;
  }

  /**
   * Store flags in cache
   * @param keyOrFlags - Either a key string (ignored, for backwards compatibility) or flags object
   * @param maybeFlags - Flags object if first param is a key
   */
  set(
    keyOrFlags: string | Record<string, boolean>,
    maybeFlags?: Record<string, boolean>,
  ): void {
    // Support both set(flags) and set(key, flags) signatures for backwards compatibility
    const flags = typeof keyOrFlags === "string" ? maybeFlags! : keyOrFlags;
    this.cache = {
      flags,
      timestamp: Date.now(),
    };
    this.persistToStorage();
  }

  /**
   * Check if cache has fresh data
   */
  hasFresh(): boolean {
    if (!this.cache) return false;
    return Date.now() - this.cache.timestamp < this.config.ttl;
  }

  /**
   * Check if cache has any data (fresh or stale)
   */
  hasAny(): boolean {
    if (!this.cache) return false;
    return Date.now() - this.cache.timestamp < this.config.staleTtl;
  }

  /**
   * Clear cached data
   */
  clear(): void {
    this.cache = null;
    if (this.config.storageKey && typeof localStorage !== "undefined") {
      try {
        localStorage.removeItem(this.config.storageKey);
      } catch {
        // Ignore storage errors
      }
    }
  }

  /**
   * Get cache statistics
   */
  getStats(): CacheStats {
    return { ...this.stats };
  }

  /**
   * Get hit rate
   */
  getHitRate(): number {
    const total = this.stats.hits + this.stats.staleHits + this.stats.misses;
    if (total === 0) return 0;
    return (this.stats.hits + this.stats.staleHits) / total;
  }

  /**
   * Load cache from localStorage
   */
  load(): boolean {
    if (!this.config.storageKey || typeof localStorage === "undefined") {
      return false;
    }

    try {
      const data = localStorage.getItem(this.config.storageKey);
      if (!data) return false;

      const entry = JSON.parse(data) as CacheEntry;

      // Only restore if within staleTtl
      if (Date.now() - entry.timestamp < this.config.staleTtl) {
        this.cache = entry;
        return true;
      }
    } catch {
      // Ignore parse errors
    }

    return false;
  }

  /**
   * Persist cache to localStorage
   */
  private persistToStorage(): void {
    if (
      !this.config.storageKey ||
      typeof localStorage === "undefined" ||
      !this.cache
    ) {
      return;
    }

    try {
      localStorage.setItem(this.config.storageKey, JSON.stringify(this.cache));
    } catch {
      // Ignore storage errors
    }
  }

  /**
   * Close the cache and persist data
   */
  close(): void {
    this.persistToStorage();
    this.removeAllListeners();
  }
}
